test/framework/src/main/java/org/elasticsearch/index/mapper/FieldTypeTestCase.java

```diff
@@ -40,4 +40,48 @@ static QueryShardContext createMockQueryShardContext(boolean allowExpensiveQueri
         return queryShardContext;
     }
 
+    protected boolean hasConfigurableDocValues() {
+        return true;
+    }
+
+    protected abstract MappedFieldType createDefaultFieldType();
+
+    protected MappedFieldType createFieldTypeWithDocValuesEnabled() {
+        throw new UnsupportedOperationException();
+    }
+
+    protected MappedFieldType createFieldTypeWithDocValuesDisabled() {
+        throw new UnsupportedOperationException();
+    }
+
+    protected boolean isAggregatableWhenDocValuesAreDisabled() {
+        return false;
+    }
+
+    protected boolean isAggregatableWhenDocValuesAreEnabled() {
+        return true;
+    }
+
```

Comment:
I can see how this association may go away, but it exists today as most of the times a field is aggregatable through doc_values, hence its exists query goes also through doc_values. The test is not so beautiful, yet it would have caught this inconsistency in the first place? I would not want to break the exists query because hasDocValue is not properly set and I thought today one way to verify that it's set correctly is by checking that a field is aggregatable, besides exceptions.

Reply:
Right, but the test structure I outlined [here](https://github.com/elastic/elasticsearch/pull/57607#discussion_r490801452) would also catch that, I think, and wouldn't involve adding all these extra methods?

Reply:
> besides exceptions

This is the bit I worry about :)